Lots of systems!

SlowFaller opened the pull request:
Fixed Scenester:
- Scenester now handles loading, pausing and unpausing, and has the stub-in for controller-reconnect.
- Scenester now correctly sets Brainiac's game state correctly
- Scenester now correctly tracks game scenes

Added new Systems!
- Added Brainiac, the autoload that handles gamepads, input events, and maintains a game state machine
- Added Critters, the autoload that handles a persistent debug menu. It can show or hide itself (key is LeftQuote or Tilde). It has a clickable menu bar that can launch debug commands or customize the debug screen. Otherwise, it is a click-through menu. It shows FPS, current scene, and the current game state. It also has a signal hookup to display anything that gets logged by Elephant.
- Added Elephant, the autoload that handles data logging, saving, and loading. It is set up to be able to save logs, config files, and game save files. It is working in a primitive state and correctly reports Keyboard presses to the Critters debug menu.

JakeGilla-DL committed:
Temp UI for displaying and modifying material data

SlowFaller committed:
Lots o stuff

Added god tithing logic and extended MaterialStack to RTithe which holds favor limits, the type of material god expects, and how much. Added cat and crow god images and set them to the god resources. Reorganized the gods folder structure. Deleted unneeded pig god scene. Changed the annoying buzzing menu ui sound to the much nicer bloop.

dasrkrain committed:
end of day event card - not done, but it's working

dasrkrain committed:
Merge branch '_working' of https://github.com/PlayDieRepeat/gmtk_2024 into _working

JakeGilla-DL committed:
Simple Art changes and additions

JakeGilla-DL committed:
Too big of a commit.  should have scoped this work better.

SlowFaller committed:
Hooked God node up to the World

Modified the world node to include debug UI labels to show tithes and favor to and from the god. Fixed a bug in the RGod data template init. World controller now precaches all materials for tithing purposes and is hooked up to God favor changing signal.

SlowFaller committed:
Coin flip

There is a coin flip now to determine one of two materials to tithe.

SlowFaller committed:
Moved pause menu to its own scene

JakeGilla-DL committed:
Scene Reconciliation - Moves everything to the level1 scene.  Refactor of the world controller to support the new structure.

SlowFaller committed:
Lots of new systems

Scenester now correctly manages loading, pausing, unpausing, and removing any unneeded scenes. Moved creation and mgmt logic into Scenester. Scenester now correctly waits for the loading screen to finish all animations before swapping to the new scene. Scenester now correctly calls Brainiac's game-state setters. Added Brainiac, which checks for input and maintains a game state machine. Added Critters to maintain a debug menu. Hooked up FPS counter, game state, and current scene info into it, along with anything that is logged to Elephant. Created Elephant, to handle saving and loading of config files, log files, and game save files. Elephant also acts as a debug logger.

SlowFaller committed:
Merge branch '_working' of https://github.com/PlayDieRepeat/gmtk_2024 into _working

SlowFaller committed:
Fixed some bugs

The game is relatively bug free.

JakeGilla-DL commented on project.godot:
```diff
@@ -10,7 +10,7 @@ config_version=5
 [application]
 
-config/name="godot_template"
+config/name="hungry_gods"
```
Thank you!

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+@export var footer_center_label: RichTextLabel
+@export var footer_right_label: RichTextLabel
+@export_group("Alpha Options")
+@export_range(0.0, 1.0, 0.1) var main_alpha: float = 0.5
```
Good use of range option

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+@export var footer_right_label: RichTextLabel
+@export_group("Alpha Options")
+@export_range(0.0, 1.0, 0.1) var main_alpha: float = 0.5
+@export_range(0.0, 1.0, 0.1) var header_alpha: float = 1.0
```
Static typing can be inferred for floats

SlowFaller replied:
I made these implicit.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+extends Node
```
no class name

SlowFaller replied:
I believe there is a conflict with giving autoloads a class name. Also it is superfluous as an autoload will have a globally accessible unique name.

SlowFaller replied:
Moved this to Brainiac. Critters is no longer an autoload, and will only load if we are in a devbuild.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+				false, false, false,
+				false, false, false]
+	_set_default_system()
+	$UI.visible = false
```
What is UI in this context?

SlowFaller replied:
A CanvasLayer object renamed to UI, I know, I was using exported vars for everything, then I toss this in, but it was the quickest way to get it running, so there is some cleanup here for sure.

SlowFaller replied:
Ok, changed it to use an exported var.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+	Scenester.scene_has_changed.connect(on_scene_has_changed)
+
+func _fps_has_changed(p_fps: int) -> void:
+	var _fps_string := footer_left_text % p_fps
```
Seems like this could be a 1-liner

SlowFaller replied:
1-lined. Now uses the built in Performance singleton.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+	footer_left_label.text = _fps_string
+
+func on_state_has_changed(p_state: String) -> void:
+	var _state_string := footer_center_text % p_state
```
Another potential 1-liner

SlowFaller replied:
You mean move it to an anon lambda function? I can do that.

SlowFaller replied:
No need for a lambda, I just removed the local variable and assigned it directly.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+	footer_center_label.text = _state_string
+
+func on_scene_has_changed(p_scene: String) -> void:
+	var _scene_string := footer_right_text % p_scene
```
Another potential 1 liner

SlowFaller replied:
No need for a lambda, I just removed the local variable and assigned it directly.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+	grid_labels[0].text = p_event
+
+func _process(delta: float) -> void:
+	var temp_time = time + delta
```
It seems like this is calculating frame rate? 
1. Do we need actually need to calculate FPS or can we get it from the engine?
2. If we do need to calculate it manually, making calculate fps function and calling it from process would be more explanatory.
3. Are there other instances than the ddebug tools where we want to output FPS? If so maybe braniac should actually own this functions and critter just poles for it.

SlowFaller replied:
I didn't check if there is an in-engine equivalent, but I will check and if there is, use it. If we go with bespoke, I see no conflict with moving it to Brainiac.

SlowFaller replied:
Ok, I hooked it up to the Performance built-in.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
```diff
@@ -0,0 +1,104 @@
+	frames += 1
+	check_grid_panel_visibility()
+
+func check_grid_panel_visibility() -> void:
```
I see a couple things about grid both bools and panels and I cant tell from code what they do are.  This makes me think a couple of possibilities
- The name isn't descriptive enough.
- There's another data structure hidden here that should be defined instead
- This is a case where an exaplanatory comment would be valuable.

SlowFaller replied:
Yeah this was sloppy. Besides a header with an interactable MenuBar, and the footer, the whole debug screen is chopped up into a 9-grid. Each grid position gets its own PanelContainer, and RichTextBox. The og idea was that you could hook up to a separate object's property getter, or maybe connect to a signal, and it would display a variable as a string. This would allow for debug and metrics to coexist and show anything that we would want to track. It was unclear how that mechanism would work, so I just made the panels and hooked up the first one to Brainac when a Key is caught as pressed. I didn't have time to think of how to code this in a data driven way, so I just made two arrays, one to hold the panels or labels, and another of the same size that would hold bools that would act as flags. This was supposed to allow a panel to get shown if it was "registered" otherwise it would hide the blank panels.

I'm not sure how useful it would be, and if we are using IMGUI then we prob. Want to cease development.

SlowFaller replied:
Added some comments, and I changed the bool array variable name to better describe its use.

JakeGilla-DL commented on scripts/_autoloads/critters.gd:
Im not sure I See anything in Critters that convinces me it should be an autoload, it seems to me it could just be instanced and removed from the tree via the input manager.  Is there something fundamental Im missing here?

SlowFaller replied:
Nope, other than making it easier for us when using our game template. More of a "works-out-of-the-box" mentality. Its fine to have an autoload just make it a child of itself if we need it. Elephant does a check to see if we are in a devbuild, and if we are, will launch the logger feature, it can also enable Critters if that is the case, otherwise, we don't want the debug menu.

SlowFaller replied:
Moved it to Brainiac for now, but running into a bug that is keeping Critters from accessing anything in its own UI, not sure why.

SlowFaller replied:
Fixed it. I was shadowing a class level variable with a new one in local method scope.

JakeGilla-DL reviewed (changes requested):
I only got through critters so far but IlL come back and hit elephant next.  theres a lot here, but so far this is some of the cleanest GodotScript I've seen you write to date.  Most my comments are more questions that descriptive actions, so the changes requested here might just be responding to the questions.

SlowFaller committed:
PR Changes and some bug fixes.

Fix a timer name bug that existed between level1 and levelTest. Now they both correctly reference a WorldTimer. Some PR changes were made, including moving Critters off of autoloads and into Brainiac. Currently showing the debug menu will result in a crash due to some weirdness on how I am instantiating it.

SlowFaller committed:
Debug Menu now works correctly.

Fixed two variables that were being shadowed by globals, aka needed to remove the "var" from the assignment in the second instance.

SlowFaller committed:
Trying to get this PR closed

Added comments and changed the variable name used to keep track of active debug panels in the GridContainer. Removed some extra lines in Critters that were not needed.

SlowFaller committed:
Minor Refactor

Changed the func _get_fps to a one-liner.

SlowFaller commented:
All items addressed.

SlowFaller committed:
Sundry items

Elephant will now set the game to restart on exit if it detects the game is running on the web.

SlowFaller committed:
Organizing our scenes and scripts.

Lots of folder creation, and moving scenes and scripts to be as close to parity as would allow. Our menus and systems root folders were getting disorganized.

SlowFaller committed:
Building out InputMap logic

Stubbed out some essential functions for input map modification. Removed system_info scene and script since that functionality is now in Elephant.

SlowFaller committed:
Renamed scripts\resources to scripts\data_templates

Having the scripts\resources folder the same as the root resources folder was confusing the hell out of me when in the file tree. We need good semantics to help with readability and clarity.

SlowFaller committed:
Revert "Renamed scripts\resources to scripts\data_templates"

This reverts commit 8d5303efff336e69b0668c5b96de8d597b2075b0.

SlowFaller committed:
Revert "Building out InputMap logic"

This reverts commit 3e3c364d86743cd2e5ee0497dcc607a608a44c0c.

SlowFaller committed:
Revert "Organizing our scenes and scripts."

This reverts commit 2880ce1d4594493c40eb5e064361a5fb6457abdf.

SlowFaller committed:
Revert "Sundry items"

This reverts commit 697c3f1ee3f3dadfbc1da3befd43e5f12331a826.

JakeGilla-DL commented on scripts/_autoloads/elephant.gd:
```diff
@@ -0,0 +1,137 @@
+	if check_dir_exists(save_game_constants.save_log_dir_path) == false:
+		var state: Error = DirAccess.make_dir_absolute(save_game_constants.save_log_dir_path)
+		if state != OK:
+			assert(state == OK, "There was a problem trying to make a directory!!")
```
Do we want to assert on being unable to create log file, or is an error/warning acceptable here?

SlowFaller replied:
Hmm, well this is assuming we are running a devbuild and we want the logger to start, so I would say yes an assert is appropriate here.

JakeGilla-DL commented on scripts/_autoloads/scenester.gd:
```diff
@@ -5,69 +5,141 @@ extends Node
+@export var reconnect_rscene: RScene
+@export var initial_rscene: RScene
 @export var should_load_initial_alt: bool = false
+@export var pause_menu: PackedScene = null
```
Why do we have a pause menu Packed scena and a pause scene r_scene which should include a packed scene.

SlowFaller replied:
Nice catch. Old way of doing things, forgot to remove that line. Its removed.

JakeGilla-DL commented on scripts/_autoloads/scenester.gd:
This class is just doing far too much at this point.  Its probably best to finish letting you cook with the input manager before we split this up. but I'm gonna create card for this.

SlowFaller replied:
How dare you.

JakeGilla-DL commented on scripts/_resources/r_game_save.gd:
Why are all these paths defined in the save game file , I'd expecte them to be in the save game constants/

SlowFaller replied:
Changed the file name from r_save_game.gd to r_save_constants.gd, the only saved instance of this data template is still called save_constants.tres.

JakeGilla-DL commented on scripts/_resources/r_god.gd:
```diff
@@ -1,14 +1,14 @@
-@export var favor: RMaterialStack = null
-@export var tithe: RMaterialStack = null
+@export var tithe: RTithe = null
+@export var favor_cap: int = 0
```
I dont think favor cap should be in the god.  If it really should, then it should be a materialstack rather than a float to allow us to couple it to the actual material.

SlowFaller replied:
I changed it to be in a resourced called [GOD_NAME}_favor_cap.tres as a MaterialStack. I also changed the one generic favor material to a cat-specific material and create a crow and pig one as well, and set them as the materials with their respective favor caps. I am loving this Resource composition stuff.

JakeGilla-DL commented on scripts/_resources/r_scene.gd:
We should probably just build an inheritance structure for scenes than adding tyhese additional rules. game scens, menu scenes, etc.  WIll make a card, shouldn't impact this review.

JakeGilla-DL commented on scripts/_resources/r_tithe.gd:
```diff
@@ -0,0 +1,12 @@
+
+class_name RTithe
+
+@export var favor_per_tithe_min: int = 0
```
All of these sshould use resource stacks instead int values to couple the appropriate material w/ the value.

SlowFaller replied:
I created r_favor_rewards.gd to house a material, low-, mid-, and high-tithing favor-reward amounts, and an awarded favor cap. 
![image](https://github.com/user-attachments/assets/6046ad75-37e5-40cb-b734-b540bc5ab6f6)

JakeGilla-DL commented on scripts/gods/god.gd:
```diff
@@ -0,0 +1,28 @@
+class_name God
+extends Node2D
```
Should inherit from Node rather than Node 2D

SlowFaller replied:
Changed.

JakeGilla-DL commented on scripts/gods/god.gd:
```diff
@@ -0,0 +1,28 @@
+signal favor_cap_reached
+
+@export var god_data: RGod
+var favor: int = 0
```
Favor should not exist on the god, should access via the material warehouse.

SlowFaller replied:
It can live on the warehouse or on the god, I don't think it matter much to me, but if it makes more sense to you for it to be in the warehouse, I can move it.

JakeGilla-DL replied:
This hasnt been updated but going to pass it through and I can refactor as a part of my changes.

JakeGilla-DL commented on scripts/systems/braniac.gd:
I trust you on this one, not sure I have much to contribute here.

JakeGilla-DL commented on scripts/systems/loading_screen.gd:
```diff
@@ -1,14 +1,26 @@
 extends CanvasLayer
```
Missing Class name

SlowFaller replied:
Named.

SlowFaller committed:
PR Latest Changes Requested

changes to God and Favor tracking

SlowFaller committed:
Favor cleanup

Removed older favor_cap resources and added in a favor_reward resource for the pig and crow gods.

SlowFaller commented:
Another commit to address changes requested. And a commit after that to cleanup some favor_cap resources that are not needed.

JakeGilla-DL commented on scripts/menus/main_menu.gd:
```diff
@@ -18,7 +18,7 @@ var button_click_sfx_instance: PooledAudioStreamPlayer
 	# We reserve sfx player instances so that we can play these sounds back monophonically
 	button_hover_sfx_instance = SoundManager.instance("ui", "ui_bloop", "UI Sounds")
-	button_click_sfx_instance = SoundManager.instance("ui", "ui_buzz", "UI Sounds")
+	button_click_sfx_instance = SoundManager.instance("ui", "ui_bloop", "UI Sounds")
```
hover and click really shouldnt be the same sound from a ux perspective. not a blocking change though.
